### PR TITLE
fix(concurrent-cdk): Move the grouping of concurrent and synchronous streams into the read and discover commands instead of when initializing the source

### PR DESCRIPTION
**Pull request by brianjlai**
## Problem

We've seen a number of issues crop up due to the fact that in low-code sources, we instantiate the streams within the source's `__init__()` method. It also created issues for config migrations because we would instantiate a source's streams using an unmigrated config which might fail validations earlier than originally expected

## Solution

This PR moves the logic of grouping and creating streams back into the `read()` and `discover()` where errors will properly be surfaced where the used to be prior to moving to the concurrent CDK.

todo :

- [x] unit tests passing
- [x] retest locally on asana or other manifest-only source with an invalid config which would not properly surface check failed message

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of concurrent and synchronous streams in data processing.
	- Simplified state management in the `ConcurrentDeclarativeSource` class.

- **Bug Fixes**
	- Improved test coverage for JSON and JSONL decoders with larger datasets.

- **Tests**
	- Updated tests for `ConcurrentDeclarativeSource` to align with new stream grouping logic.
	- Increased line count in test responses to better simulate larger datasets.

### CODE REVIEW

File: airbyte_cdk/sources/declarative/concurrent_declarative_source.py [hunk 1]
```diff
@@ -86,23 +86,10 @@ def __init__(
             component_factory=component_factory,
         )
 
+        # todo: We could remove state from initialization. Now that streams are grouped during the read(), a source
+        #  no longer needs to store the original incoming state. But maybe there's an edge case?
```
Comment (brianjlai) [inline airbyte_cdk/sources/declarative/concurrent_declarative_source.py:90]: added this comment as a todo, i don't think we use state anywhere, but in essence of fewer changes to fix the bug.

File: unit_tests/sources/declarative/decoders/test_json_decoder.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def test_jsonl_decoder(requests_mock, response_body, expected_json):
 def large_event_response_fixture():
     data = {"email": "email1@example.com"}
     jsonl_string = f"{json.dumps(data)}\n"
-    lines_in_response = 2  # ≈ 58 MB of response
+    lines_in_response = 2_000_000  # ≈ 58 MB of response
```
Comment (brianjlai) [inline unit_tests/sources/declarative/decoders/test_json_decoder.py:57]: i might've accidentally changed this by accident in a past review

File: unit_tests/sources/declarative/test_concurrent_declarative_source.py [hunk 3]
```diff
@@ -866,7 +867,21 @@ def _mock_party_members_skills_requests(http_mocker: HttpMocker) -> None:
...
+def mocked_init(self, is_sequential_state: bool = True):
+    """
+    This method is used to patch the existing __init__() function and always set is_sequential_state to
+    false. This is required because we want to test the concurrent state format. And because streams are
+    created under the hood of the read/discover/check command, we have no way of setting the field without
+    patching __init__()
+    """
+    self._is_sequential_state = False
+
+
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply unit_tests/sources/declarative/test_concurrent_declarative_source.py:879]: sir this is a ~wendy's~ test already

File: unit_tests/sources/declarative/decoders/test_json_decoder.py (hunk 2 above)
Comment (pnilan) [reply unit_tests/sources/declarative/decoders/test_json_decoder.py:57]: Ha I remember that PR, too.

**Review by pnilan — APPROVED** [review]
Looks good!

Tested it with `source-sendgrid`. Resolved the issue.

**Comment by brianjlai** [comment]
https://github.com/airbytehq/airbyte/actions/runs/12188006089/job/34000075439

regression test counts were successful and expected diffs verified.

**Pull request merged**